### ISSUE TITLE
[Feature Request] Solution for storing large intermediate results in DB

### ISSUE THREAD

**Issue by bartvanb** [vantage6/vantage6#268]
**Problem description**
For some algorithms (especially Machine Learning/Deep Learning), the output of an algorithm might be very large (in the order of GBs). This would increase the database size quickly, especially if multiple intermediate results or multiple tasks are stored. Also, it would lead to very slow API calls if you want to retrieve such results.

**Desired solution**
It would be better to store these results outside the database. For instance, if we simply store them as files, we could refer in the database to the file paths to retrieve the results. 
Note that if we facilitate working with very large data, we should also consider how to clean up. Perhaps there should be an (optional) automated cleanup after X time.

**Comment by tim-hendriks** [comment]
I'd like to work on this feature within the context of current DHD projects, where the preference is to use Azure blob storage to store the intermediate results. I've proposed the following solution, please let me know if you think it makes sense for the Vantage6 V4 codebase (and of course I'd like to make it compatible with V5 if possible) and where we can improve.

I've opted to fix the currently encountered database issue first before looking into possible API issues. The assumption (let me know if I'm wrong in this) is that storing intermediate results in this way will be required regardless of how API problems are resolved later on.

# Large Object Storage Solution Proposal
To allow for the storage of large results outside of the existing relational database for the Vantage6 server, the option for a large object store will be added to the Vantage6 server codebase. This large object store functionality should be extensible to work for any kind of large object store.

## Configuration

The option to add a large_result_store value is added to the configuration. The field is not present in the default server configuration, and in the case the value is not set, vantage6 utilizes a simple file-based storage for the results. The possible values of large_result_store can be expanded to include different types of object stores, such as S3 compatible stores. Initially, we start implementation with Azure Blob Storage. If large_result_store is used, of course additional configuration such as connection parameters and the name for the storage account will also need to be provided.

## Implementation

![Image](https://github.com/user-attachments/assets/6b8a4baf-1aea-4998-8754-ab14b691d051)

The result storage adapter is a simple interface with methods to get, store and delete results from a large object store. These methods will use Run Model objects as in- and outputs. In the case of Azure blob storage, task-ids from the Run object will be used for Azure storage container-ids.
UUIDs will be generated and stored as pointers in the original Run object. These pointers are then used by ResultStorageAdapter to retrieve the associated objects when required. To prevent issues in case of misconfiguration or allow upgrading of an existing V6 Server, a field will be added to the Run object to signify which object store should be used to retrieve the result if any.

**Comment by bartvanb** [comment]
Nice to see your plans regarding this! I'll try to respond where necessary. Note that I haven't looked at the details of how these storage interfaces work so I might be lacking some knowledge there.

> I've opted to fix the currently encountered database issue first before looking into possible API issues. The assumption (let me know if I'm wrong in this) is that storing intermediate results in this way will be required regardless of how API problems are resolved later on.

I'm not entirely sure I understood this. Am I correct in assuming that you mean the following: you first want to get the route to store data in blob storage working. This will probably create the need to adapt the API, which you plan to do after that. I think that approach makes sense: the first challenge is to get the data transfer from the node working, and then the API can be adjusted.

### General comments

I think the solution you propose to handle everything at the server makes sense as you don't need to configure anything at the nodes - they just send the data and how the server stores it, is up to the server. It basically means we hardly (if at all) update the nodes, which is convenient.

I am wondering though how big the burden of doing this on the server would be: every result is passed from the node to the server, and then from the server to the storage (right?). This probably means the server will be caught up for a while in downloading/uploading data. How do you ensure the server is not blocked during this time? Would it be possible directly upload the data from the node to the storage or am I overseeing a reason that makes that infeasible? (maybe the server could provide the node with an address to upload the result to sending the algorithm run details?)

I'm also wondering how this works with the clients. When downloading large data, the request will probably take some time. In Python client (and any other programming-language-based client), this should not be a big issue, but for the UI it would currently mean the page will basically be loading for a while, which is not nice. But I guess this is already the case at the moment.

### v5 compatibility

Just wanted to note on this for future reference. I think most of the data communication solution would be compatible with v5, as it's mostly changing API fields with more options. I think the result storage adapter might change a bit more, but I'm not entirely sure what is necessary to implement there for v4. In v5, based on a quick search, I think you would specify a volume like [here](https://stackoverflow.com/questions/59263656/accessing-azure-file-share-from-local-kubernetes-cluster) or [here](https://medium.com/@mayurguptacs59/using-azure-file-share-as-persistence-volumes-with-kubernetes-029638dc9015). K8s supports storage types listed [here](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#types-of-persistent-volumes) which includes azure. Maybe this will enable generalizing that class.

**Comment by tim-hendriks** [comment]
Apologies for not getting to this sooner. You understood correctly indeed, I'm hoping that this way of storing references to the actual results is required in any scenario we end up with. So if you agree with the approach I can start with that already.

As far as the API work I think we share a lot of the concerns you mention. I don't have experience scaling flask, but from what I gather we should be able to scale workers to allow for multiple concurrent requests.

I'd expect the burden on the server from the much larger results used in the future to be a bigger issue. In theory we should be able to stream these results from storage through the server to the client without storing or keeping in memory the entire result on the server. If I understand correctly we should also be able to utilize the RabbitMQ solution to scale more server instances if needed. Whether this will be manageable/desirable in the current V6 Server is difficult for me to say at this point. Perhaps another future bottleneck to keep in mind is the aggregator node if it needs to deal with very large results, if it is pulling all these large results and aggregating them.

On loading data the UI being problematic, my assumption was that in the case of large data we don't send the results to the UI. If I understood our previous conversation well this is not required or desirable right?

In the earlier design I assumed streaming data through the server. As you mention the other approach would be to communicate directly from node to storage. Some of the pros and cons for each:

## Through the Server

### Pros

- No extra authentication and authorization work, more control over this
- No extra configuration and implementation on the Node side for different storage solutions (S3, Azure etc.)
- No need to create service accounts for each participating site
- No need for extra firewall configuration / allowlisting
- Potentially possible to use a basic file storage solution for simple usecases while holding results in the central server 

### Cons

- Difficult to say what the current potential bottlenecks in the server will be and how much work to fix them is required
- May not be the most future-proof solution once result sizes become extremely large
- Costs incurred due to more resources required for the server(s)

## Straight to storage

The best description and approximation of what this solution will look like (reading and writing directly from storage) has been worked out by DHD in a pdf file. Since I'm not a 100% sure if it is okay to share I will ask Yannick to share below.

### Pros

- More efficient for extremely large results
- Less development work (?)

### Cons

- Each time a node is added service accounts need to be created first for the appropriate storage
- Extra firewall configuration / allowlisting
- Vantage6 based permissions will have to be 'synced' with the different storage solutions (keycloak may provide solutions here but I think that will be V5 and we need this in V4)
- Depending on the implementation, changing to a different storage type or location might mean changing node configuration for all sites individually
- More blob storage data transfer related costs
- Requires extra configuration and perhaps discussions with the sites (?) to allow an extra entity

Please feel free to add and correct me where I'm wrong (Yannick also, I may be exaggerating some of these site configuration issues). I'm also omitting any encryption related repercussions if there are any.

**Comment by bartvanb** [comment]
Nice overview of the pros and cons. If I can abstract it a bit, the question comes down to if we prefer to go via the direct route node-storage that requires more configuration for data parties, or via the server, which may result in bottlenecks if server is not scaled properly.
I guess I need to think on it a little more but I'm leaning toward the solution via the server now: it is probable that the node configuration needs to be changed for such storage settings sometimes, which is not easy for big networks. The server bottlenecks is an issue that seems more solveable. Indeed, horizontal scaling of the server can be done if you set up RabbitMQ. 
I'm not exactly sure how the streaming of files through the server would go, but if we can circumvent blocking the API with a long request, that makes a bottleneck less likely.

**Comment by frankcorneliusmartin** [comment]
> As far as the API work I think we share a lot of the concerns you mention. I don't have experience scaling flask, but from what I gather we should be able to scale workers to allow for multiple concurrent requests.

Yes the vantage6 allows for scaling, however do not forget that the sessions between clients - server and between node - server are sticky (always connect to the same instance). This then can still lead to bottlenecks when request are taking long to process.

An idea: why not defined a second API next to the server which handles uploads (non sticky API)

**Comment by yannick-vinkesteijn** [comment]
We also investigated a design where Blob Storage would be accessed directly by the nodes instead of via the central server. The full design proposal is attached:
[azure-cloud-architecture.pdf](https://github.com/user-attachments/files/19535459/azure-cloud-architecture.pdf)

## Current issue
Large files are currently stored as JSON in a PostgreSQL database. This leads to performance issues and file size limitations when handling large files such as trained model parameters.

## Proposal
Use Blob Storage as an external file store for large files generated by a Vantage6 node, and return only a reference to that file in the task result.

## Implementation
- **Azure service principals** Each Vantage6 organization has its own Azure Storage container and Service Principal, isolating access by design.
- **Terraform automation** Used to manage storage containers, service principals, and permissions (handled outside the Vantage6 scope).
- **Fixed central node**  Since the central node needs access to data from other nodes, it must be a fixed node with persistent (elevated) access to the storage containers of all collaborating organizations.

**Note:** This design could potentially be simplified by avoiding hospital-specific containers and instead using collaboration-specific containers with read-only and write-only access policies.

To add to the existing pros and cons list:

## Pros
- No need to stream large files through the central server.
- The central server remains lightweight and avoids handling large payloads directly.
- Per organization storage access control by default.
- Concurrency is handled by Blob Storage; scaling depends only on blob storage limits.
- Supports any file type to blob storage through direct file streaming, this allows for more efficient compressed file types such as Parquet or [safetensors](https://huggingface.co/docs/safetensors/index). 

## Cons
- Each node requires cloud access setup, including IP whitelisting. (For AIOC whitelisting is already implemented at hospitals) -> already mentioned by @tim-hendriks 
- Adds cloud infrastructure complexity and orchestration overhead for setting up and managing terraform.
- Requires additional setup to ensure vantage6 permissions are reflected in Azure.
- A fixed central node adds to setup complexity and reduces Vantage6 session flexibility (only needed when collaboration nodes are not allowed to access each other's data).
- Server side retrieval for monitoring tasks needs to be implemented (however this part seems to cause some problems so not sure if this is going to stay?)

@tim-hendriks I'm not sure about the cloud costs increasing with the node solution. Won't the same amount of data be transferred for both solutions? Allowing for compressed file types might reduce cloud costs.

**Comment by tim-hendriks** [comment]
You're right Yannick, the cloud costs argument doesn't make any sense, costs are the same there in all 3 options. 

The advantage of the hybrid solution would indeed be that FastAPI is probably more suited to this than Flask. Disadvantage is then we still have to consider how to deal with authentication and authorization on this second API in a v4 that doesn't include Keycloak right? I'd be interested in the current temporary solution indeed.

I've tested a simple flask app for 20 concurrent requests on a 20GB mock parquet file which seemed to work well, though the test is not 100% representative. We discussed yesterday I would further investigate what the concurrency and large data issues are. I will share a PoC repository here with the flask app and further tests on the V6 infra.
